### DISCUSSION TITLE
Skip pre-loading Prometheus metric names - let the user type a few characters first

### DISCUSSION THREAD

**Discussion by danstadler-pdx** [grafana/grafana#63063] [Data Sources]
# Summary

A Grafana customer has been using the [Google Cloud Monitoring datasource](https://grafana.com/docs/grafana/latest/datasources/google-cloud-monitoring/). For most use cases it is great, but when using the “Service” pulldown to select Prometheus metrics, this customer experiences very slow load times, to the point where the browser starts asking if the user would like to wait, or kill the page.

This customer has a moderate number of unique Prometheus metrics (under 2000) - however this doesn’t seem like an unreasonable number for a customer to have; it’s reasonable to think that the UI should be able to handle this volume of unique metrics.

This FR is suggesting a change in the Explore area of the Grafana UI that could help in this situation.

## Feature summary is as follows:

Allow customers to add a configuration value that instructs the Explore section to offer a choice to users using the GCM datasource. The choice would be between “auto-load all values” and “don’t auto-load all values”. The user would specify which GCM services they want to see this choice for, for example:

[explore]
GCM_services_skip_autoloading_metric_names = [“Prometheus”]

If a Service is included in that list, then the following behavior could happen.

### Implementation Idea 1

User goes to GCM datasource in Explore

User selects Prometheus as a service

Metric Name field does not auto-populate. Instead, an open text field is presented next to it.

User types a number of characters into the field

Upon reaching a minimum number of characters (perhaps also configurable) the auto-search turns on, and populates the pulldown.

If the user wants to force a search before reaching the minimum character count, there could be a button to do this, perhaps with a modal “are you sure” before performing the name search.

The user might be offered some further control, like setting their own minimum character count before auto-search starts.

### Implementation Idea 2

If a user is in Explore and using a GCM datasource, there could be a choice between “Prometheus” and “all other Services”. 

When choosing Prometheus, the UI could switch out and be replaced by a Prometheus-specific UI, which implements the above ideas in Implementation Idea 1. 

Calling this out as a separate idea, because it might make the Prometheus-specific functionality more re-usable for other datasources which also talk to Prometheus back-ends.

**Comment by Elfo404** [comment]
What are the Explore specifics here? to me this really looks like a performance issue with the GCM datasource and nothing really related to explore other than the fact that the customer is experiencing this in Explore (the same performance issues would most likely happen when using that editor when creating a dashboard).

I moved the discussion to the appropriate section, and i'd also ping @grafana/partner-plugins as this may be considered a performance bug on the plugin itself

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82787. Feel free to continue the discussion around this there. Thank you!
